회원정보 수정 기능 개발

Habist opened the pull request:
## 작업 내용
- 회원정보 수정 요청에 해당하는 도메인 클래스를 추가했습니다.
- 인증 서비스에 인증된 사용자 아이디 사용을 위한 메소드를 추가했습니다.
- 사용자 서비스에 사용자 아이디와 변경 요청을 기반으로 사용자 정보를 수정하는 메소드를 추가했습니다.
- 사용자 아이디에 해당하는 정보를 변경하는 쿼리를 추가했습니다.

## 주의 사항

>  참조이슈
> #14

Habist committed:
feat: [#14] 회원정보 수정 기능 개발

회원정보 수정 요청에 해당하는 도메인 클래스를 추가했습니다.
인증 서비스에 인증된 사용자 아이디 사용을 위한 메소드를 추가했습니다.
사용자 서비스에 사용자 아이디와 변경 요청을 기반으로 사용자 정보를 수정하는 메소드를 추가했습니다.
사용자 아이디에 해당하는 정보를 변경하는 쿼리를 추가했습니다.

f-lab-dev commented on src/main/java/com/campool/controller/UserController.java:
```diff
@@ -36,4 +37,9 @@ public void logoutUser() {
         authService.deauthenticate();
     }
 
+    @PostMapping("/users/update")
```
이 문서를 읽어보시고 수정해주시면 좋을 것 같네요~

https://docs.microsoft.com/ko-kr/azure/architecture/best-practices/api-design

Habist replied:
#13 리뷰의 답변처럼 로그인 검증 기능을 인터셉터로 구현하면서 발생한 문제를 해결하기 위해 또 다른 문제를 만들었네요

리뷰 반영하겠습니다~

f-lab-dev commented on src/main/java/com/campool/controller/UserController.java:
```diff
@@ -36,4 +37,9 @@ public void logoutUser() {
 
+    @PostMapping("/users/update")
+    public void updateUser(@Valid UserUpdateRequest userUpdateRequest) {
+        userService.updateById(authService.getUserId(), userUpdateRequest);
```
`authService.getUserId()` 이것도 매번 호출해서 가져오기보다는 `@PathVariable` 처럼 argument로 주입받아보면 좋을 것 같네요~

Habist replied:
리뷰 반영했습니다!

f-lab-dev commented on src/main/java/com/campool/service/AuthService.java:
```diff
@@ -10,4 +10,6 @@
 
     boolean isValidAuthentication();
 
+    String getUserId();
```
이름에서 `현재 인증되어있는 유저의 아이디를 가져온다`라는게 명확하게 보이면 좋을 것 같아요~

Habist replied:
좀 더 명확한 의미를 가지도록 변경하였습니다!

Habist committed:
improve: [#14 - 리뷰 반영] REST API 디자인 준수 및 인증 서비스 메소드명 수정

회원정보 수정하는 요청 URL을 REST API 디자인에 더 맞게 준수하는 쪽으로 변경하였습니다.
인증 서비스 메소드 중 인증된 사용자 아이디를 가져오는 메소드명을 좀 더 명확한 의미로 변경하였습니다.

f-lab-dev commented on src/main/java/com/campool/controller/UserController.java:
```diff
@@ -37,9 +38,9 @@ public void logoutUser() {
-    public void updateUser(@Valid UserUpdateRequest userUpdateRequest) {
-        userService.updateById(authService.getUserId(), userUpdateRequest);
+    @PostMapping("/users/{id}")
+    public void updateUser(@Valid UserUpdateRequest userUpdateRequest, @PathVariable String id) {
```
아 path variable은 비유를 한 거였고, 이것처럼 현재 로그인한 유저의 아이디를 자동으로 이렇게 주입해줄 수 있지 않을까를 여쭤본거였습니다~

Habist replied:
현재 세션에 저장되어 있는 사용자 아이디를 주입받을 수 있도록 명시하는 애노테이션과 주입을 하는 AOP를 추가하였습니다~

Habist committed:
feat: [#14 - 리뷰 반영] 현재 로그인 유저 아이디 주입 애노테이션, AOP 추가

현재 세션에 저장되어 있는 사용자 아이디를 주입받을 수 있도록 명시하는 애노테이션을 추가하였습니다.
현재 세션에 저장되어 있는 사용자 아이디를 주입해주는 AOP를 추가하였습니다.

f-lab-dev commented on src/main/java/com/campool/aop/PresentUserIdResolver.java:
```diff
@@ -0,0 +1,41 @@
+        Method method = ((MethodSignature) joinPoint.getSignature()).getMethod();
+        Annotation[][] parameterAnnotations = method.getParameterAnnotations();
+        Class<?>[] parameterTypes = method.getParameterTypes();
+        loop:
```
for문이 1개인데 라벨을 달아주는건 불필요한 구문추가가 아닐까요?

Habist replied:
```
for (int i = 0; i < args.length; i++) {
    for (Annotation annotation : parameterAnnotations[i]) {
        if (annotation.annotationType() == PresentUserId.class
                && parameterTypes[i] == String.class) {
            args[i] = authService.getAuthenticatedUserId();
            break loop;
        }
    }
}
```
Argument 개수 만큼 반복 돌리고 Annotation 2차원 배열도 반복하는 중첩반복아닌가요??
매개변수가 여러 개이고 각 변수 별로 선언된 애노테이션이 많으면 충분히 비효율적인 반복이 진행되는 상황이라 생각했습니다.

f-lab-dev reviewed (commented):
좋네요~ 이제 다시 한번 개선을 해보자면 스프링에서는 이런 기능을 `HandlerMethodArgumentResolver`라는 것으로 제공하고 있습니다. 그것도 한번 알아보시면 좋을 것 같아요~

Habist committed:
improve: [#14 - 리뷰 반영] 현재 사용자 아이디 주입하는 AOP Spring Resolver 기반으로 변경

현재 세션에 저장되어 있는 사용자 아이디를 주입해주는 AOP를 제거하고
Spring HandlerMethodArgumentResolver 기반으로 변경하였습니다.

Habist commented:
AOP기반에서 HandlerMethodArgumentResolver 기반으로 바꿔봤습니다!

f-lab-dev commented on src/main/java/com/campool/resolver/PresentUserIdResolver.java:
```diff
@@ -0,0 +1,32 @@
+    @Override
+    public boolean supportsParameter(MethodParameter parameter) {
+        return (parameter.getParameterAnnotation(PresentUserId.class) != null)
+                && parameter.getParameterType().equals(String.class);
```
만약 개발자의 실수로 String 타입이 아닌 타입으로 받게될 경우 `null`이 들어가게 될 것 같은데 어노테이션을 사용하는 개발자의 입장에서는 `왜 어노테이션 달았는데 null이 들어오지?`하고 본인의 실수를 인지하지 못하지 않을까요?

타입이 맞지 않으면 예외를 던지게 해주는게 더 실수를 알아차리기 좋을 것 같네요~

Habist replied:
넵! String 타입이 아닌 경우에는 메시지와 예외를 발생하도록 변경하였습니다!

Habist committed:
improve: [#14 - 리뷰 반영] PresentUserIdResolver String 타입 확인 로직 추가

타입 확인을 위해 String 이외의 타입이 들어오면 예외를 발생하도록 변경하였습니다.

Habist committed:
fix: [#14] PresentUserId 파라미터 타입 체크 로직 버그 수정

애노테이션이 없는 매개 변수에도 타입 체크를 하는 버그를 수정하였습니다.
-> PresentUserId 애노테이션이 있는 파라미터만 String 타입 체크를 하도록 변경

Habist committed:
Merge branch 'feature/13' of https://github.com/f-lab-edu/campool into feature/14

 Conflicts:
	src/main/java/com/campool/configuration/WebConfiguration.java
	src/main/java/com/campool/controller/UserController.java
